### ISSUE TITLE
Add elastic demand and demand side response example

### ISSUE THREAD

**Issue by pz-max** [PyPSA/PyPSA#574]
## Describe the feature you'd like to see
It seems demand side considerations plays around the world an increasing role.
I think we could add a few examples illustrating how:
- A. elastic demand and
- B. demand side response
can be modeled within PyPSA.

**Requirements:**
A, as indicated below, can be solved iteratively. What we only need is to create an arbitrary elastic demand curve and show how it impacts the results compared to a inelastic case.
![image](https://user-images.githubusercontent.com/61968949/221236183-a8128111-fb73-4691-a2a6-1f13a62faa7d.png)

B. Is already well used in PyPSA-Eur/PyPSA-Earth and implemented as negative "load" generator. We just need to show how it works in a minimal example.

**Comment by Irieo** [comment]
@pz-max @FabianHofmann elastic demand can be modeled in one-step optimization. For that one needs (i) to rewrite objective to social welfare optimization and (ii) represent price via affine term [a+b*x], where X is supplies of good in a given node, time and [a,b] are demand functions intercept and slope that are assumed (typically from econometric literature). 
technically, we can do it asap mighty Linopy can handle QP problems.

**Comment by pz-max** [comment]
> @pz-max @FabianHofmann elastic demand can be modeled in one-step optimization. For that one needs (i) to rewrite objective to social welfare optimization and (ii) represent price via affine term [a+b*x], where X is supplies of good in a given node, time and [a,b] are demand functions intercept and slope that are assumed (typically from econometric literature). technically, we can do it asap mighty Linopy can handle QP problems.

Really cool :)
- So my suggestions works as LP problem & with LP solvers
- Your suggestions requires Linopy to support QP problems (issue is [here](https://github.com/PyPSA/linopy/issues/18)), requires to test PyPSA with QP problems, requires QP solvers (usally are 1.5-2 times slower than LP solvers as you estimated in a personal chat)

Modelling elastic demand with QP problems is definetly the long-term better option! Would be cool to have examples in PyPSA for both.

**Comment by Irieo** [comment]
The key problem for studies with elastic demand is parametrization. We can do econometrics to estimate the demand elasticity to price; however, in 2030/40/50, with all the electrification of demand, EVs, smart meters and digitization, the elasticity of demand will likely be much higher.

**Comment by coroa** [comment]
Is there literature on the estimation of that elasticity already?

**Comment by Irieo** [comment]
@coroa yes, there are many studies. For example, [Labandeira et al. 2017](https://doi.org/10.1016/j.enpol.2017.01.002) survey 428 papers that use various econometric techniques to evaluate energy demand elasticity to price. They also groupby analyses per energy carrier, as well as short-/long-run estimates. For electricity, they find average short-run price elasticity of electricity demand at 0.13.

**Comment by Irieo** [comment]
> @coroa yes, there are many studies. For example, [Labandeira et al. 2017](https://doi.org/10.1016/j.enpol.2017.01.002) survey 428 papers that use various econometric techniques to evaluate energy demand elasticity to price. They also groupby analyses per energy carrier, as well as short-/long-run estimates. For electricity, they find average short-run price elasticity of electricity demand at 0.13.

-0.13, obv.

**Comment by davide-f** [comment]
As an additional (temporart) option may be ti implement N load shedding at incremental prices and quantity.
E.g. let's say that we have 10MW, and let's say that for everu additional MW to shed shed energy becomes 1€/kWh more costly. 1€/kWh for the first MW shed, then 2, then 3 etc.
Then, an approximation (step-wise) of that linear demand curve can be done implementing N (fake) generators similarly to the current load shedding but with different max capacities (in this case all 1MW) and different marginal costs (from 1 up to 10 theoretically).

Personally, I like very much the introduction of quadratic objective as I've seen generators modelled with that function.

**Comment by nworbmot** [comment]
Here is a filthy hack for putting quadratic terms for generators when using `linopt` (i.e. `n.lopf(pyomo=False)`) when using gurobi:
```
def run_and_read_gurobi():
....
    if "quadratic_coefficient" in n.generators:
        print("Adding quadratic coefficients")
        to_add = []
        for gen in n.generators.index:
            qc = n.generators.at[gen,"quadratic_coefficient"]
            if qc == 0:
                continue
            print(f"gen has non-zero QC of {qc}")
            for s,i in n.vars["Generator"]["pnl"]["p"][gen].items():
                x = m.getVarByName(f"x{i}")
                to_add.append((qc*n.snapshot_weightings.at[s,"generators"],x))
        linear_objective = m.getObjective()
        new_objective = linear_objective + sum(qc*x*x for qc,x in to_add)
        m.setObjective(new_objective)
```
And then in code for elastic demand (for utility 2100*d -10d**2 with inverse demand 2100 - 20*d):
```
        network.add("Generator","load",
                    bus="electricity",
                    carrier="load",
                    marginal_cost=2100,
                    p_max_pu=0,
                    p_min_pu=-1,
                    p_nom=assumptions["load"])
        network.generators["quadratic_coefficient"] = 0.                             
        network.generators.at["load","quadratic_coefficient"] = 10.
``
I think this functionality would be good to add for all components (links, stores, etc.), since all can have quadratic cost functions.
It has zero effect on barrier solving time for gurobi (may even make it faster).
You need quadratic_coefficient >= 0. obviously to stay convex.

**Comment by fneum** [comment]
With the [new PyPSA release](https://github.com/PyPSA/PyPSA/releases/tag/v0.24.0) you can now have a quadratic cost coefficient for all components.

```py
n.add("Generator",
    "generator",
    marginal_cost=50,
    marginal_cost_quadratic=0.5,
    ...
)
```
